openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+///          4) Allow for the grid types of the sdf and ext to be different
+///
+/// @note    Solves the (simplified) Eikonal Eq: @f$|\nabla \phi|^2 = 1@f$ and
+///          performs velocity extension,  @f$\nable f\nabla \phi = 0@f$, both
+///          by means of the fast sweeping algorithm detailed in:
+///          "A Fast Sweeping Method For Eikonal Equations"
+///          by H. Zhao, Mathematics of Computation, Vol 74(230), pp 603-627, 2004
+///
+/// @details The algorithm used below for parallel fast sweeping was first publised in:
+///          "New Algorithm for Sparse and Parallel Fast Sweeping: Efficient
+///          Computation of Sparse Distance Fields" by K. Museth, ACM SIGGRAPH Talk,
+///          2017, http://www.museth.org/Ken/Publications_files/Museth_SIG17.pdf
+
+#ifndef OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+
+//#define BENCHMARK_FAST_SWEEPING
+
+#include <type_traits>// for static_assert
+#include <cmath>
+#include <limits>
+#include <thread>
+
+#include <tbb/blocked_range.h>
+#include <tbb/parallel_for.h>
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_invoke.h>
+#include <tbb/enumerable_thread_specific.h>
+
+#include <openvdb/math/Math.h> // for Abs() and isExactlyEqual()
+#include <openvdb/math/Stencils.h> // for GradStencil
+#include <openvdb/util/PagedArray.h>
+#include <openvdb/tree/LeafManager.h>
+#include "LevelSetUtil.h"
+#include "Morphology.h"
+
+#include "Statistics.h"
+#ifdef BENCHMARK_FAST_SWEEPING
+#include <openvdb/util/CpuTimer.h>
+#endif
+
+#include <openvdb/math/Stats.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+/// @brief Converts a scalar fog volume into a signed distance function. Active input voxels
+///        with scalar values above the given isoValue will have NEGATIVE distance
+///        values on output, i.e. they are assumed to be INSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note Strictly speaking a fog volume is normalized to the range [0,1] but this
+///       method accepts a scalar volume with an arbitary range, as long as the it
+///       includes the @a isoValue.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
```

Comment:
This does not sound like an exceptional situation.  The result of that should just be an empty SDF grid?

Reply:
when attempting to create an SDF from a volume at a non-existing iso-value, I think it's misleading to return an empty grid. First, an empty grid does exist unless of course, you mean a shared point to NULL? If not, what should be background value be (recall it's supposed to be an SDF)? Returning a null smart pointer is not much differet from throwing .. is it.

Reply:
I think these routines should all maintain the grid topology.  So return the original grid with the same activity mask; but all the values set to the outside background value.

Reply:
The throw has been removed and instead empty shared pointers are returned. If would be east to instead return empty grids (with the provided background values), if you still prefer that?

Reply:
I'd prefer empty grids to null pointers.  Indeed, I think from my understanding they should be empty grids with the same topology as the input, but all active voxels set to the background value, because these functions are not supposed to affect topology?

What happens if a disconnected input grid has one connected component have no iso-surface crossing?  Does it properly still extend the components that have a crossing?

Reply:
The problem is, the background value of an SDF has a very specific meaning, which is breaks down for empty grids. The background value for a truncated SDF is the maximum positive distance value, i.e. the value at the out-most level-set of the narrow band. However, when the fundamental assumptions of the fast sweeping algorithm as not satisfied (e.g. no boundary conditions), then obviously not distance values can be computed, which in turn means it's unclear what the background value should be of the "empty sdf with active voxels" that you want. I could set it to std::numerical_limits<T>::max() but why is that useful?